### ISSUE TITLE
Port listening / IP interface

### ISSUE THREAD

**Issue by spyjo** [anthonyraymond/joal#44]
Hello,

I have installed joal on my dedicated server, and I have some suggests:

- "Listening for incoming peer connections on port XXXX" => since I have a firewall on my server, it would be great to be able to choose this port manually in the configuration file, so I can open it in the firewall.

- "Fetching ip from: http://ident.me/" , "Successfully fetch public IP address:  XXXXX", "Ip reported to tracker will be XXXXX" => I have multiple IP addresses on my server, and in my case for my real Transmission server I have fixed one specific IP address. It would be great to be able to specify also the same IP address in joal configuration file, so others ips are not connected with torrents, and the seed would come from the same IP than the real client.

Thank you for your work :)

**Comment by anthonyraymond** [comment]
Ok, this shouldn't be too hard to implement. I'll do it asap, but i'm pretty busy these days.

**Comment by anthonyraymond** [comment]
Hi @spyjo 

Just to make it clear what you want.

For the second point, you want to force using a specific network interface or you want to mock the ip address?
Not sure why you would like to use a second network interface (unless you have two different internet access point with different public ips). And mocking the ip adress won't make the trick since the tcp packets will still contains your real IP (which will be different from the mocked one).

**Comment by anthonyraymond** [comment]
I can force the listening port, but i wont be able to force the network interface, there are no way to do such things on java.
But you can probably do that on the operating system level. There is a windows software to do that. There is probably a way to achieve this on linux as well

**Comment by spyjo** [comment]
in fetchIp() in ConnectionHandler.java, may be you can use an "force public IP" option to force a choosen ip instead of using the IP returned by one of the provider ? Since it seems that the IP returned by this function is used to be reported to the torrent tracker

**Comment by anthonyraymond** [comment]
Yeah this is definely possible, but i believe this does not makes sense.
I can force http request to include a fake IP address in the request parameter `ip` (note that not all client sends this parameter, and most of the times trackers ignore it anyway). But i can't spoof your IP address in TCP/UDP packets, nor i can force java to root his traffic though a specific network interace. This is something that has to be done on the os level.

Plus such a feature will be confusing for users, if you can choose a specific IP you'd expect it to be the ip address that will appear on the tracker, and this is not the case.

**Comment by spyjo** [comment]
> nor i can force java to root his traffic though a specific network interace

It's doable almost everywhere, I would be surprised if Java doesn't have this option too.
For a listening daemon, you can bind to a specific IP address.
For remote connections, there must be some option to force the socket to connect from a specific source.

**Comment by anthonyraymond** [comment]
Some ways may exists if i would try to dig in the code for every http, tcp and udp request that the program will ever made. And with the condition that every library that i use to do such request allow me to change the network interface. And such things are error prone for future development.
This is way too much if for such a feature that can most likely be solved at the os level in a more elegant way (IMHO).

https://stackoverflow.com/questions/49989230/setting-jvm-local-network-interface

**Comment by deadpansociety** [comment]
It'd be helpful to have customisable port number
